packages/amplify-graphql-relational-transformer/src/schema.ts

```diff
@@ -197,115 +293,177 @@ export const ensureHasManyConnectionField = (
     return;
   }
 
-  const connectionAttributeName = getConnectionAttributeName(object.name.value, field.name.value);
+  const sortKeyFields = getSortKeyFields(ctx, object);
+
+  const primaryKeyField = getObjectPrimaryKey(object);
+  const connectionFieldName = primaryKeyField.name.value;
+  config.connectionFields.push(
+    connectionFieldName,
+    ...sortKeyFields.map(it => getSortKeyConnectionAttributeName(ctx.featureFlags, object.name.value, field.name.value, it.name.value)),
+  );
 
   const relatedTypeObject = ctx.output.getType(relatedType.name.value) as ObjectTypeDefinitionNode;
+  const connectionAttributeName = getConnectionAttributeName(ctx.featureFlags, object.name.value, field.name.value, connectionFieldName);
+
+  const primaryKeyConnectionFieldType = ctx.featureFlags.getBoolean('useFieldNameForPrimaryKeyConnectionField') ? getBaseType(primaryKeyField.type) : 'ID';
+
   if (relatedTypeObject) {
-    ctx.output.putType(updateTypeWithConnectionField(relatedTypeObject, connectionAttributeName, isNonNullType(field.type)));
+    const updatedFields = [...relatedTypeObject.fields!];
+    updatedFields.push(
+      ...getTypeFieldsWithConnectionField(updatedFields, connectionAttributeName, primaryKeyConnectionFieldType, isNonNullType(field.type)),
+    );
+    sortKeyFields.forEach(it => {
+      updatedFields.push(...getTypeFieldsWithConnectionField(updatedFields,
+        getSortKeyConnectionAttributeName(ctx.featureFlags, object.name.value, field.name.value, it.name.value),
+        getBaseType(it.type),
+        isNonNullType(field.type)));
+    });
+    ctx.output.putType({
+      ...relatedTypeObject,
+      fields: updatedFields,
+    });
```

Comment:
Ditto, this looks quite similar to 331-348 and 355-372, and can probably be refactored into a nicely-named helper.